lib/internal/worker.js

```diff
@@ -97,12 +99,34 @@ class Worker extends EventEmitter {
       }
       argv = options.argv.map(String);
     }
-    if (!options.eval) {
-      if (!path.isAbsolute(filename) && !/^\.\.?[\\/]/.test(filename)) {
-        throw new ERR_WORKER_PATH(filename);
+
+    let url;
+    if (filename instanceof URL) {
+      if (filename.protocol !== 'file:') {
+        throw new ERR_WORKER_UNSUPPORTED_PROTOCOL(filename.protocol);
+      }
+      if (options.eval) {
+        throw new ERR_INVALID_ARG_VALUE(
+          'options.eval',
+          options.eval,
+          'must be false when \'filename\' is a URL object'
+        );
+      }
+      url = filename;
+      filename = fileURLToPath(url);
+    } else {
+      validateString(filename, 'filename');
+      if (!options.eval) {
+        if (!path.isAbsolute(filename) && !/^\.\.?[\\/]/.test(filename)) {
+          throw new ERR_WORKER_PATH(filename);
+        }
+        filename = path.resolve(filename);
       }
-      filename = path.resolve(filename);
 
+      url = options.eval ? null : pathToFileURL(filename);
+    }
```

Comment:
Reuse https://github.com/nodejs/node/blob/884e287199901853a30781a0b398e6c3e4a07eb1/lib/internal/fs/utils.js#L577-L581 for consistency (e.g. checks, errors thrown) with `fs`?